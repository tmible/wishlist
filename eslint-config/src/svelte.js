import svelte from "eslint-plugin-svelte";
import globals from "globals";

export default [
  ...svelte.configs["flat/recommended"],
  {
    languageOptions: {
      globals: {
        ...globals.browser,
        ...globals.node,
      },
      ecmaVersion: "latest",
      sourceType: "module",
      parserOptions: {
        extraFileExtensions: [ ".svelte" ],
      },
    },

    settings: {
      "import/parsers": {
        "svelte-eslint-parser": [ ".svelte" ],
        "espree": [ ".js" ],
      },
    },

    rules: {
      /* import */
      // svelte does tree shaking when compiling
      "import/no-namespace": "off",

      /* svelte */
      "svelte/infinite-reactive-loop": "error",
      "svelte/no-dom-manipulating": "error",
      "svelte/no-dupe-on-directives": "error",
      "svelte/no-dupe-use-directives": "error",
      "svelte/no-reactive-reassign": "error",
      "svelte/no-store-async": "error",
      "svelte/require-store-callbacks-use-set-param": "error",
      "svelte/block-lang": "error",
      "svelte/no-ignored-unsubscribe": "error",
      "svelte/no-immutable-reactive-statements": "error",
      "svelte/no-inline-styles": "error",
      "svelte/no-reactive-functions": "error",
      "svelte/no-reactive-literals": "error",
      "svelte/no-useless-mustaches": "error",
      "svelte/require-each-key": "error",
      "svelte/require-event-dispatcher-types": "error",
      "svelte/require-stores-init": "error",
      "svelte/valid-each-key": "error",
      "svelte/derived-has-same-inputs-outputs": "error",
      "svelte/first-attribute-linebreak": "error",
      "svelte/html-closing-bracket-new-line": "error",
      "svelte/html-closing-bracket-spacing": "error",
      "svelte/html-quotes": "error",
      "svelte/html-self-closing": [
        "error",
        {
          "void": "never",
          "normal": "never",
          "svg": "always",
          "math": "never",
          "component": "always",
          "svelte": "always",
        },
      ],
      "svelte/indent": [ "error", { ignoredNodes: [ "ConditionalExpression" ] } ],
      "svelte/max-attributes-per-line": [ "error", { multiline: 1, singleline: 100 } ],
      "svelte/mustache-spacing": "error",
      "svelte/no-extra-reactive-curlies": "error",
      "svelte/no-spaces-around-equal-signs-in-attribute": "error",
      "svelte/prefer-class-directive": "error",
      "svelte/prefer-style-directive": "error",
      "svelte/shorthand-attribute": "error",
      "svelte/sort-attributes": [
        "error",
        {
          order: [
            // `this` property.
            "this",
            // `bind:this` directive.
            "bind:this",
            // `id` attribute.
            "id",
            // `name` attribute.
            "name",
            // `slot` attribute.
            "slot",
            // `--style-props`
            { match: "/^--/u", sort: "ignore" },
            // `style` attribute, and `style:` directives.
            [ "style", "/^style:/u" ],
            // `class` attribute.
            "class",
            // `class:` directives.
            { match: "/^class:/u", sort: "ignore" },
            // other attributes.
            {
              match: [ "!/:/u", "!/^(?:this|id|name|style|class)$/u", "!/^--/u" ],
              sort: "ignore",
            },
            // `bind:` directives (other then `bind:this`), and `on:` directives.
            [ "/^bind:/u", "!bind:this", "/^on:/u" ],
            // `use:` directives.
            { match: "/^use:/u", sort: "ignore" },
            // `transition:` directive.
            { match: "/^transition:/u", sort: "ignore" },
            // `in:` directive.
            { match: "/^in:/u", sort: "ignore" },
            // `out:` directive.
            { match: "/^out:/u", sort: "ignore" },
            // `animate:` directive.
            { match: "/^animate:/u", sort: "ignore" },
            // `let:` directives.
            { match: "/^let:/u", sort: "ignore" },
          ],
        },
      ],
      "svelte/spaced-html-comment": "error",
      "svelte/no-inner-declarations": "error",
      "svelte/no-trailing-spaces": "error",
    },
  },
  {
    files: [ "**/*.svelte" ],
    rules: {
      // svelte components can import themselves
      "import/no-self-import": "off",
    },
  },
  {
    files: [ "**/**.test.js", "**/**.test.svelte.js", "**/__tests__/**" ],
    settings: { "import/ignore": [ "vitest", "@testing-library/svelte" ] },
  },
];
