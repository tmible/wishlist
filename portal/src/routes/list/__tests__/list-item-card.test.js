// @vitest-environment jsdom
import { cleanup, render, screen } from '@testing-library/svelte';
import { userEvent } from '@testing-library/user-event';
import { afterEach, describe, expect, it, vi } from 'vitest';
import ListItemCard from '../list-item-card.svelte';

vi.mock(
  '$lib/components/telegram-entities/parser.svelte',
  async () => ({ default: await import('./mock.svelte').then((module) => module.default) }),
);
vi.mock(
  '../list-item-form.svelte',
  async () => ({ default: await import('./mock.svelte').then((module) => module.default) }),
);

describe('list item card', () => {
  afterEach(() => {
    vi.clearAllMocks();
    cleanup();
  });

  it('should be displayed as skeleton', () => {
    const { container } = render(ListItemCard, { isSkeleton: true });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should be displayed default', () => {
    const { container } = render(
      ListItemCard,
      { listItem: { name: 'name', category: { id: 0, name: 'category' } } },
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should be displayed for reordering', () => {
    const { container } = render(
      ListItemCard,
      { listItem: { name: 'name' }, isReorderModeOn: true },
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should enable editing mode', async () => {
    const { container } = render(
      ListItemCard,
      { listItem: { name: 'name', category: { id: 0, name: 'category' } } },
    );
    const user = userEvent.setup();
    await user.click(screen.getByText(/Редактировать/, { selector: 'button' }));
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should request item deletion', async () => {
    const dispatch = vi.fn();
    vi.resetModules();
    vi.doMock('svelte');
    vi.doMock(
      '$lib/components/telegram-entities/parser.svelte',
      async () => ({ default: await import('./mock.svelte').then((module) => module.default) }),
    );
    vi.doMock(
      '../list-item-form.svelte',
      async () => ({ default: await import('./mock.svelte').then((module) => module.default) }),
    );
    vi.mocked(
      await import('svelte').then(({ createEventDispatcher }) => createEventDispatcher),
    ).mockReturnValue(dispatch);
    const listItem = { id: 'id', name: 'name', category: { id: 0, name: 'category' } };
    render(
      await import('../list-item-card.svelte').then((module) => module.default),
      { listItem },
    );
    const user = userEvent.setup();
    await user.click(screen.getByText(/Удалить/, { selector: 'button' }));
    expect(dispatch).toHaveBeenCalledWith('delete', listItem);
  });
});
