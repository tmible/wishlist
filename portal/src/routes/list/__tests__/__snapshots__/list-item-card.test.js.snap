// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`list item card > should be displayed as skeleton 1`] = `
<div
  class="card bg-base-100 md:shadow-xl"
>
  <div
    class="card-body prose"
  >
    <h2
      class="skeleton card-title"
    >
       
    </h2>
     
    <p>
      <span
        class="skeleton mb-px block"
      >
         
      </span>
       
      <span
        class="skeleton mb-px block"
      >
         
      </span>
       
      <span
        class="skeleton mb-px block"
      >
         
      </span>
       
      <span
        class="skeleton mb-px block"
      >
         
      </span>
    </p>
     
    <div
      class="card-actions"
    >
      <button
        class="skeleton btn btn-neutral w-full md:flex-1"
      >
         
      </button>
       
      <button
        class="skeleton btn btn-neutral w-full md:flex-1"
      >
         
      </button>
    </div>
  </div>
</div>
`;

exports[`list item card > should be displayed default 1`] = `
<div
  class="card bg-base-100 md:shadow-xl"
>
  <div
    class="card-body prose"
  >
    <h2
      class="card-title"
    >
      name
    </h2>
     
    <div
      class="prose"
    >
       
      

Mock
    </div>
     
    <div
      class="card-actions pt-4"
    >
      <button
        class="btn btn-outline btn-neutral w-full md:flex-1"
      >
        <svg
          class="lucide-icon lucide lucide-pencil"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M21.174 6.812a1 1 0 0 0-3.986-3.987L3.842 16.174a2 2 0 0 0-.5.83l-1.321 4.352a.5.5 0 0 0 .623.622l4.353-1.32a2 2 0 0 0 .83-.497z"
          />
          
          <path
            d="m15 5 4 4"
          />
          
          
        </svg>
        
            Редактировать
      </button>
       
      <button
        class="btn btn-outline btn-error w-full md:flex-1"
      >
        <svg
          class="lucide-icon lucide lucide-trash-2"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3 6h18"
          />
          
          <path
            d="M19 6v14c0 1-1 2-2 2H7c-1 0-2-1-2-2V6"
          />
          
          <path
            d="M8 6V4c0-1 1-2 2-2h4c1 0 2 1 2 2v2"
          />
          
          <line
            x1="10"
            x2="10"
            y1="11"
            y2="17"
          />
          
          <line
            x1="14"
            x2="14"
            y1="11"
            y2="17"
          />
          
          
        </svg>
        
            Удалить
      </button>
    </div>
  </div>
</div>
`;

exports[`list item card > should enable editing mode 1`] = `
<div
  class="card bg-base-100 md:shadow-xl"
>
  <div
    class="card-body prose"
  >
     
    

Mock
  </div>
</div>
`;
